### PR TITLE
Support in-cloud transition, cross-cloud transition and expiration of…

### PR DESCRIPTION
**Pull request by sfzeng**
**What this PR does / why we need it**:
This PR is to provide the ability of scheduling and data migration for lifecycle management. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
This PR is to fixes [issue694](https://github.com/opensds/multi-cloud/issues/694)

**Special notes for your reviewer**:
With this PR, object expiration and transition both in a specific backend or cross different backends can are supported.
But as the multipart upload is not finished, abandon data left by multi-part upload is not support to be cleaned by lifecycle management.
And garbage clean is also left, which is traced by [issue697](https://github.com/opensds/multi-cloud/issues/697) 

**Release note**:

```release-note
```

### CODE REVIEW

File: s3/pkg/service/object.go [hunk 1]
```diff
@@ -363,6 +364,255 @@ func (s *s3Service) GetObject(ctx context.Context, req *pb.GetObjectInput, strea
...
+		err = s.MetaStorage.AddGcobjRecord(ctx, newObj)
+		if err != nil {
+			log.Errorf("failed to add gcobj record[%v], err:%v", newObj, err)
+			// if delete failed, no error return, because gc will clean it
+			ierr := s.MetaStorage.DeleteGcobjRecord(ctx, newObj)
+			log.Infof("delete source object from gc finished, err:", ierr)
+			return err
+		}
+		// step 2: copy data
+		err = s.copyData(ctx, srcSd, targetSd, srcObject.Object, targetObject)
```
Comment (hopkings2008) [inline s3/pkg/service/object.go]: after the copy, it is better to check the etag(md5, if has) with the original one to make sure the data integrity.

File: s3/pkg/service/object.go [hunk 2]
```diff
@@ -363,6 +364,255 @@ func (s *s3Service) GetObject(ctx context.Context, req *pb.GetObjectInput, strea
...
+		if err != nil {
+			log.Errorln("failed to create storage. err:", err)
+			return err
+		}
+
+		// steps of moving object data: add target object for gc -> copy data -> update meta(remove target object from
+		// gc, update object to be the target one, and add source object for gc in a transaction) -> delete source
+		// object data from backend-> delete source object from gc. If crash happened, gc service will clean data.
+		// step 1: add new object for gc
+		newObj := &meta.Object{Object: targetObject}
```
Comment (hopkings2008) [inline s3/pkg/service/object.go]: before copy, the object id of target is the same as original one, and after the copy, the object id of the target is changed. so, before copy, the target object id is saved in gc, and if it is fail to copy, then we will clean the original object?

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go [hunk 3]
```diff
@@ -158,14 +164,15 @@ func handleBucketLifecyle(bucket string, rules []*osdss3.LifecycleRule) error {
 			inRules = append(inRules, &v)
 		}
 
-		if rule.AbortIncompleteMultipartUpload.DaysAfterInitiation > 0 {
+		// Currently, incomplete multipart upload data will be cleaned by GC, so no lifecycle rules for those data
```
[the remark this answers is not in the document]
Comment (sfzeng) [reply dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: Thanks it's done.

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go [hunk 4]
```diff
@@ -179,13 +186,14 @@ func handleBucketLifecyle(bucket string, rules []*osdss3.LifecycleRule) error {
 	// End: Log for debug
 	schedSortedActionsRules(&inRules)
 
-	sort.Stable(abortRules)
+	/*sort.Stable(abortRules)
```
[the remark this answers is not in the document]
Comment (sfzeng) [reply dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: Thanks it's done.

File: datamover/pkg/drivers/lifecycle/crosscloud_transition.go [hunk 5]
```diff
@@ -17,65 +17,58 @@ package lifecycle
...
-	log.Infof("part_size=%d, err=%v.\n", part_size, err)
-	if err == nil {
-		// part_size must be more than 5M and less than 100M
-		if part_size >= 5 && part_size <= 100 {
-			PART_SIZE = part_size * 1024 * 1024
-			log.Infof("Set PART_SIZE to be %d.\n", PART_SIZE)
-		}
+func copyObj(obj *osdss3.Object, targetLoc *LocationInfo) error {
+	log.Infof("copy object[%s], size=%d\n", obj.ObjectKey, obj.Size)
+	if obj.Size <= 0 {
```
[the remark this answers is not in the document]
Comment (sfzeng) [reply datamover/pkg/drivers/lifecycle/crosscloud_transition.go]: Thanks, I just removed this check, please review again.

File: s3/pkg/datastore/aws/aws.go [hunk 6]
```diff
@@ -83,25 +85,43 @@ func (ad *AwsAdapter) Put(ctx context.Context, stream io.Reader, object *pb.Obje
...
-	log.Infof("put object[AWS S3] begin, objectId:%s\n", objectId)
-	_, err := uploader.Upload(&s3manager.UploadInput{
-		Bucket: &bucket,
-		Key:    &objectId,
-		Body:   stream,
-		// Currently, only support STANDARD for PUT.
-		StorageClass: aws.String(constants.StorageClassAWSStandard),
-	})
-	log.Infof("put object[AWS S3] end, objectId:%s\n", objectId)
+	d, err := ioutil.ReadAll(stream)
```
[the remark this answers is not in the document]
Comment (sfzeng) [reply s3/pkg/datastore/aws/aws.go]: Thanks for your suggestion, it's fixed, please review again.

File: s3/pkg/service/object.go (hunk 2 above)
Comment (sfzeng) [reply s3/pkg/service/object.go]: No, check UpdateMetaAfterCopy please.

File: s3/pkg/datastore/ceph/ceph.go [hunk 7]
```diff
@@ -57,7 +48,7 @@ func (ad *CephAdapter) Put(ctx context.Context, stream io.Reader, object *pb.Obj
 	cephObject := bucket.NewObject(bucketName)
 	d, err := ioutil.ReadAll(stream)
```
[the remark this answers is not in the document]
Comment (sfzeng) [reply s3/pkg/datastore/ceph/ceph.go]: Same as above.

File: s3/pkg/service/object.go (hunk 1 above)
Comment (sfzeng) [reply s3/pkg/service/object.go]: Thanks for your suggestion, it's fixed, please review again.

**Comment by hopkings2008** [comment]
LGTM

File: s3/pkg/conf/tidb.sql [hunk 8]
```diff
@@ -196,6 +196,29 @@ CREATE TABLE IF NOT EXISTS `objects` (
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;
 /*!40101 SET character_set_client = @saved_cs_client */;
 
+
+--
+-- Table structure for table `gcobjs`
+--
+/*!40101 SET @saved_cs_client     = @@character_set_client */;
+/*!40101 SET character_set_client = utf8 */;
+CREATE TABLE IF NOT EXISTS `gcobjs` (
```
[the remark this answers is not in the document]
Comment (sfzeng) [reply s3/pkg/conf/tidb.sql]: etag and tier is removed, please check.

[the remark this answers is not in the document]
Comment (sfzeng) [reply s3/pkg/conf/tidb.sql]: Thanks for your suggestion, I removed etag and tier, please review again.

**Comment by sfzeng** [comment]
@rhsakarpos @nguptaopensds @kumarashit @sunfch Please help to review this PR, thanks a lot !

File: dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go [hunk 9]
```diff
@@ -55,7 +58,7 @@ func loadStorageClassDefinition() error {
 		return fmt.Errorf("get tier definition failed")
 	} else {
 		log.Infof("res.Transition:%v", res.Transition)
```
[the remark this answers is not in the document]
Comment (sfzeng) [reply dataflow/pkg/scheduler/lifecycle/lifecycle_sched.go]: Yes we can, as it's no affect to the feature, I'll merge them in the next PR.

**Pull request merged**